Flat buffer support for Druid-Lookups: WIP

pranavbhole opened the pull request:
<!-- The following line must be included in your pull request -->
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

patelh commented:
Maybe we should see if there is a better way to configure the schema factory.  Maybe we should create a new version here so we can cleanly move from old to new.

patelh commented:
as far as rocksdbmanager and kafkamanager are concerned, they only care about serializing and deserializing and then getting access to the data by field name and value IIRC.

panditsurabhi commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/entity/CacheActionRunner.java:
```diff
@@ -15,29 +12,39 @@
                 return lookupService.lookup(new LookupService.LookupData(extractionNamespace, key, valueColumn.get(), decodeConfigOptional));
             }
 
+            if (routeToNextChain(schemaFactory)) {
```
Not sure if chain of responsibility is the right pattern here, since there's no 'chaining' of steps. 'schemaFactory' object can be type flatbuffer, protobuf or some other derived class of BaseSchemaFactory. Falling back to protobuf silently may not be the expected behavior always.

pavanab4u replied:
I second that. there is no chain of responsibility here. you should call appropriate CacheActionRunner based on the configuration in the ExtractionNamespace which is already done here https://github.com/yahoo/maha/blob/master/druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/namespace/RocksDBExtractionNamespace.java#L80-L90

pranavbhole replied:
removed the chain of responsibility.

pranavbhole replied:
and added new cache action runner.

panditsurabhi commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/entity/CacheActionRunnerFlatBuffer.java:
```diff
@@ -0,0 +1,124 @@
+
+public class CacheActionRunnerFlatBuffer implements BaseCacheActionRunnerChain {
+
+    private static final Logger LOG = new Logger(CacheActionRunner.class);
```
Let the logger be CacheActionRunnerFlatBuffer.class

pranavbhole replied:
fixed

pavanab4u commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/RocksDBLookupExtractor.java:
```diff
@@ -47,21 +45,21 @@
     public RocksDBLookupExtractor(RocksDBExtractionNamespace extractionNamespace, Map<String, U> map,
                                   LookupService lookupService, RocksDBManager rocksDBManager, KafkaManager kafkaManager,
-                                  ProtobufSchemaFactory protobufSchemaFactory, ServiceEmitter serviceEmitter) {
+                                  BaseSchemaFactory schemaFactory, ServiceEmitter serviceEmitter) {
```
how is this going to be instantiated? during query time this factory is used for deserialization. It is instantiated here https://github.com/yahoo/maha/blob/master/druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/cache/MahaNamespaceExtractionCacheManager.java#L391
where concrete class is injected https://github.com/yahoo/maha/blob/master/druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/cache/MahaNamespaceExtractionCacheManager.java#L73

pranavbhole replied:
fixed,  added the comment in the class MahaNamespaceExtractionCacheManager

patelh commented:
You should be able to specify the protobuf message class or flatbuffer message class in the json for extraction namespace.  Best if you create a new RocksDBV2ExtractionNamespace and clean up.

patelh commented:
Right now we have to provide config for protobuf schema factory in server side config.  It makes it more cumbersome to use.  Do all configuration as part of the json.  If protobuf or flatbuffer message class doesn't exist, error out.  This way all the user has to do is drop the message format jar in the libs folder, add the maha lookup extension and config and do rest via lookup registration.

pranavbhole commented:
Sorry for not addressing comments since long, had to wait for Druid Upgrade.  Now resuming the work on this, I have just rebased and added the parallel namespace for FlatBuffer extraction with optional  MahaNamespaceExtractionConfig config.

panditsurabhi commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/schema/flatbuffer/FlatBufferWrapper.java:
```diff
@@ -0,0 +1,26 @@
+    public abstract FlatBufferBuilder createFlatBuffer(Map<String, FlatBufferValue> nameToValueMap);
+
+    public byte[] toByteArr(ByteBuffer byteBuffer) {
+        byte[] arr = new byte[byteBuffer.remaining()];
```
Let's add a null check for byteBuffer

panditsurabhi commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/entity/CacheActionRunnerFlatBuffer.java:
```diff
@@ -0,0 +1,126 @@
+            } catch (Exception e) {
+                LOG.error(e, "Caught exception while updating cache");
+                serviceEmitter.emit(ServiceMetricEvent.builder().build(MonitoringConstants.MAHA_LOOKUP_UPDATE_CACHE_FAILURE, 1));
+                e.printStackTrace();
```
Do we need printStackTrace since e is logged above?

panditsurabhi commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/namespace/JDBCExtractionNamespace.java:
```diff
@@ -123,6 +123,11 @@ public boolean isCacheEnabled() {
     }
 
+    @Override
+    public boolean isFlatBufferNamespace() {
```
Instead of adding this method in every namespace, either:
1. Add default implementation in ExtractionNamespace interface with false
2. Better - create a method 'getNamespaceType' in ExtractionNamespace which returns appropriately for each namespace ("jdbc", "flatbuffer" etc)

pranavbhole commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/cache/MahaNamespaceExtractionCacheManager.java:
```diff
@@ -387,6 +390,8 @@ private LookupExtractor getLookupExtractor(final ExtractionNamespace extractionN
             return new JDBCLookupExtractorWithLeaderAndFollower((JDBCExtractionNamespaceWithLeaderAndFollower) extractionNamespace, map, lookupService);
         } else if (extractionNamespace instanceof JDBCExtractionNamespace) {
             return new JDBCLookupExtractor((JDBCExtractionNamespace) extractionNamespace, map, lookupService);
+        } else if (extractionNamespace instanceof RocksDBExtractionNamespace && extractionNamespace.getType() == ExtractionNameSpaceSchemaType.FlatBuffer) {
```
@pavanab4u  this is the how it instantiate the RocksDBLookupExtractor supporting flatbuffer schema.

pranavbhole replied:
I have added ExtractionNameSpaceSchemaType to all existing name spaces in order to state which schema it supports.

pranavbhole commented:
> You should be able to specify the protobuf message class or flatbuffer message class in the json for extraction namespace. Best if you create a new RocksDBV2ExtractionNamespace and clean up.

Added the config in the RocksDBExtractionNamespace for accepting flat buffer cache runner. And added an optional config into MahaNamespaceExtractionConfig to configure the name of the flat buffer schema factory class  which is dumped into jar.

pranavbhole commented:
There are not breaking changes along with this PR, all configurations are optional.

pavanab4u commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/namespace/JDBCExtractionNamespace.java:
```diff
@@ -123,6 +122,11 @@ public boolean isCacheEnabled() {
 
+    @Override
+    public ExtractionNameSpaceSchemaType getType() {
+        return ExtractionNameSpaceSchemaType.Protobuf;
```
technically JDBCExtractionNamespace is not using any schema. it is storing String as it is. you may want to have `None` as additional type for this case.

pranavbhole replied:
done

pavanab4u commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/namespace/MongoExtractionNamespace.java:
```diff
@@ -118,6 +118,11 @@ public boolean isCacheEnabled() {
 
+    @Override
+    public ExtractionNameSpaceSchemaType getType() {
+        return ExtractionNameSpaceSchemaType.Protobuf;
```
same here. you need `None` type

pranavbhole replied:
done

pavanab4u commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/entity/BaseCacheActionRunner.java:
```diff
@@ -0,0 +1,36 @@
+
+    void validateSchemaFactory(BaseSchemaFactory schemaFactory);
+
+    boolean flatBufferSupport();
```
i think this is redundant. implementation type should tell what kind of schema it supports. you can just check `instanceof`

pranavbhole replied:
removed this

pavanab4u commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/server/lookup/namespace/entity/CacheActionRunner.java:
```diff
@@ -78,14 +78,16 @@ public String handleDecode(DecodeConfig decodeConfig, Message parsedMessage, Des
             , RocksDBExtractionNamespace extractionNamespace) {
         if (extractionNamespace.isCacheEnabled()) {
             try {
+                validateSchemaFactory(schemaFactory);
```
it would be better to use bounded type parameter and restrict this class to be used only with `ProtobufSchemaFactory`. with this it will be a compile time check as against current runtime check.

pranavbhole replied:
schema factory is compile time validation now, had to add extra logic at the rocksdb namespace

pranavbhole commented on druid-lookups/src/main/java/com/yahoo/maha/maha_druid_lookups/query/lookup/namespace/ExtractionNamespace.java:
```diff
@@ -29,4 +28,6 @@
 
     boolean isCacheEnabled();
+
+    ExtractionNameSpaceSchemaType getType();
```
done
